src/subtyping.md

```diff
@@ -15,5 +15,70 @@ fn bar<'a>() {
     let t: &'a str = s;
 }
 ```
-Since `'static` "lives longer" than `'a`, `&'static str` is a subtype of
-`&'a str`.
+
+Since `'static` outlives `'a`, `&'static str` is a subtype of `&'a str`.
+
+Subtyping also exists for [higher-ranked] function pointer and trait object
+types. Replacing a higher ranked lifetime with a concrete lifetime produces a
+subtype.
+
+```rust
+fn fun_ptr() {
+    // Explicitly f: &(for<'a> fn(&'a i32) -> &'a i32).
+    let f: &(fn(&i32) -> &i32) = &((|x| x) as fn(&i32) -> &i32);
+    let g: &(fn(&'static i32) -> &'static i32) = f;
+}
+
+fn fun_trait() {
+    // Explicitly f: &(for<'a> Fn(&'a i32) -> &'a i32).
+    let f: &(Fn(&i32) -> &i32) = &(|x| x);
+    let g: &(Fn(&'static i32) -> &'static i32) = f;
+}
+```
+
+## Variance
+
+Variance is a property that generic types have with respect to their arguments.
+A generic type's *variance* in a parameter is how the subtyping of the
+parameter affects the subtyping of the type.
+
+* `F<T>` is *covariant* over `T` if `T` being a subtype of `U` implies that
+  `F<T>` is a subtype of `F<U>` (subtyping "passes through")
+* `F<T>` is *contravariant* over `T` if `T` being a subtype of `U` implies
+  that `F<U>` is a subtype of `F<T>`
+* `F<T>` is *invariant* over `T` otherwise (no subtyping relation can be derived)
+
+Variance of types is automatically determined as follows
+
+| Type                          | Variance in `'a`  | Variance in `T`   |
+|-------------------------------|-------------------|-------------------|
+| `&'a T`                       | covariant         | covariant         |
+| `&'a mut T`                   | covariant         | invariant         |
+| `*const T`                    |                   | covariant         |
+| `*mut T`                      |                   | invariant         |
+| `[T]` and `[T; n]`            |                   | covariant         |
+| `fn() -> T`                   |                   | covariant         |
+| `fn(T) -> ()`                 |                   | contravariant     |
+| `std::cell::UnsafeCell<T>`    |                   | invariant         |
+| `std::marker::PhantomData<T>` |                   | covariant         |
+| `Trait<T> + 'a`               | covariant         | invariant         |
```

Comment:
Thing I don't know but someone should check: is `&'a mut Trait<T> + 'a` actually invariant in `'a`? (this is implied my the entries for `&mut T` and `Trait<T> + 'a`, but I would almost expect the compiler to magic this away.

Reply:
It looks like it is invariant, but there is a coercion from `&mut (Trait + 'a)` to `&mut Trait + 'b` where `'a: 'b`.  The same as `&mut (Trait + Send)` and `&mut Trait`.
```rust
trait A {}
// This compiles (also with `*mut`, or `Box<RefCell>`)
fn coercion<'a, 'b>(x: &'b mut (A + 'static)) -> &'b mut (A + 'a) {
    x
}
// This does not
fn not_subtype<'a, 'b, 'c>(x: &'c &'b mut (A + 'static)) -> &'c &'b mut (A + 'a) {
    x
}
```